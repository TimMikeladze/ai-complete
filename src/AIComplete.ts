import {
  Configuration,
  CreateEditRequest,
  CreateEditResponse,
  OpenAIApi
} from 'openai'
import { globbySync, Options } from 'globby'
import { readFileSync } from 'fs'
import {
  CreateCompletionRequest,
  CreateCompletionResponse
} from 'openai/api.js'
import { ConfigurationParameters } from 'openai/configuration.js'

export interface DataInput {
  type: 'filepath' | 'text'
  value: string
}

export interface AICompleteOptions {
  openAI: {
    config: ConfigurationParameters
  }
}

export type RequestArgs = {
  data?: DataInput[]
  globby?: {
    options?: Options
    patterns: string | readonly string[]
  }
}

export interface CreateCompletionArgs {
  input: (args: {
    args: CreateCompletionArgs
    filePath: string
    text: string
  }) => Promise<{
    content?: string
    maxTokens?: number
    prompt: string
    request?: Partial<CreateCompletionRequest>
  }>
  output: (args: {
    args: RequestArgs
    content: string
    data: CreateCompletionResponse
    filePath: string
    maxTokens?: number
    prompt: string
    request: Partial<CreateCompletionRequest>
  }) => Promise<{
    choice: any
  }>
}

export interface CreateEditArgs {
  input: (args: {
    args: CreateEditArgs
    filePath: string
    text: string
  }) => Promise<{
    content?: string
    input?: string
    instruction: string
    request?: CreateEditRequest
  }>
  output: (args: {
    args: CreateEditArgs
    content: string
    data: CreateEditResponse
    filePath: string
    input?: string
    instruction: string
    request: CreateEditRequest
  }) => Promise<{
    choice: any
  }>
}

export interface CreateInsertArgs {
  input: (args: {
    args: CreateInsertArgs
    filePath: string
    text: string
  }) => Promise<{
    content?: string
    input?: string
    instruction: string
    request?: CreateEditRequest
  }>
  output: (args: {
    args: CreateEditArgs
    content: string
    data: CreateEditResponse
    filePath: string
    input?: string
    instruction: string
    request: CreateEditRequest
  }) => Promise<{
    choice: any
  }>
}

export const defaultCreateCompletionRequest = {
  temperature: 0.7,
  top_p: 1,
  frequency_penalty: 0,
  presence_penalty: 0,
  model: 'text-davinci-003'
}

export const defaultCreateEditRequest = {
  temperature: 0.7,
  top_p: 1,
  model: 'text-davinci-edit-001'
}

export class AIComplete {
  private readonly openai: OpenAIApi
  private readonly options: AICompleteOptions
  private static MAX_TOKENS = 4000

  constructor(options: AICompleteOptions) {
    const configuration = new Configuration(options.openAI.config)
    this.openai = new OpenAIApi(configuration)
    this.options = options
  }

  private async request(args: RequestArgs & {}, fn: any) {
    const data = []

    const addGlobbyInput = (
      patterns: string | readonly string[],
      options?: Options
    ) => {
      data.push(
        ...globbySync(patterns, options).map((filePath) => ({
          type: 'filepath',
          value: filePath
        }))
      )
    }

    const addFileInput = (filePath: string) => {
      data.push({
        type: 'filepath',
        value: filePath
      })
    }

    const addTextInput = (text: string) => {
      data.push({
        type: 'text',
        value: text
      })
    }

    if (args.globby) {
      addGlobbyInput(args.globby.patterns, args.globby.options)
    }

    if (args.data) {
      for (const inputItem of args.data) {
        if (inputItem.type === 'filepath') {
          addFileInput(inputItem.value)
        }
        if (inputItem.type === 'text') {
          addTextInput(inputItem.value)
        }
      }
    }
    const getContent = (inputItem): string => {
      let content: string
      if (inputItem.type === 'filepath') {
        content = readFileSync(inputItem.value, 'utf8')
      }
      if (inputItem.type === 'text') {
        content = inputItem.value
      }
      return content
    }

    const res = []

    for (const inputItem of data) {
      try {
        res.push(
          await fn(
            args,
            getContent(inputItem),
            inputItem.type === 'filepath' ? inputItem.value : undefined
          )
        )
      } catch (e) {
        console.error(e)
      }
    }

    return res
  }

  public static getTokenCount(prompt: string, _maxTokens?: number) {
    // count alphanumeric characters in prompt
    const promptLength = prompt.replace(/[^a-z0-9]/gi, '').length

    const maxTokens =
      _maxTokens || Math.min(promptLength, AIComplete.MAX_TOKENS)

    if (maxTokens > AIComplete.MAX_TOKENS) {
      throw new Error(
        `maxTokens must be less than ${AIComplete.MAX_TOKENS}. Try breaking the input into smaller chunks.`
      )
    }
    return maxTokens
  }

  public async createCompletion(args: RequestArgs & CreateCompletionArgs) {
    return this.request(args, this.createCompletionFromText.bind(this))
  }

  public async createCompletionFromText(
    args: RequestArgs & CreateCompletionArgs,
    text: string,
    filePath?: string
  ) {
    const input = await args.input({
      args,
      text,
      filePath
    })

    const prompt = `${input.prompt} ${input.content || text}`.trim()

    const maxTokens = AIComplete.getTokenCount(prompt, input.maxTokens)

    const request: CreateCompletionRequest = {
      prompt,
      max_tokens: maxTokens,
      ...defaultCreateCompletionRequest,
      ...input.request
    }

    const response = await this.openai.createCompletion(request)
    const data = response.data

    const output = await args.output({
      data,
      request,
      maxTokens,
      prompt,
      args,
      content: text,
      filePath
    })

    return {
      data,
      input,
      request,
      output,
      args,
      filePath
    }
  }

  public async createEdit(args: RequestArgs & CreateEditArgs) {
    return this.request(args, this.createEditFromText.bind(this))
  }

  public async createEditFromText(
    args: RequestArgs & CreateEditArgs,
    text: string,
    filePath?: string
  ) {
    const input = await args.input({
      args,
      text,
      filePath
    })

    const content = input.content || text
    const instruction = input.instruction

    const request: CreateEditRequest = {
      instruction,
      input: content,
      ...defaultCreateEditRequest,
      ...input.request
    }

    const response = await this.openai.createEdit(request)
    const data = response.data

    const output = await args.output({
      data,
      request,
      instruction,
      input: content,
      args,
      content: text,
      filePath
    })

    return {
      output,
      data,
      request,
      instruction,
      input: content,
      args,
      content: text,
      filePath
    }
  }
}
