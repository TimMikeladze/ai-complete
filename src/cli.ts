#!/usr/bin/env node

import { Command } from 'commander'
// import { AIComplete } from './AIComplete'

// interface CommanderOptions {}
;(async () => {
  const program = new Command()

  program.parse(process.argv)

  // const options: CommanderOptions = program.opts()

  // const translator = new AIComplete()

  // await aic.initialize();

  process.exit(0)
})()
