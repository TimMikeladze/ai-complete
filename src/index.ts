import { AIComplete } from './AIComplete'

export * from './AIComplete'

export default AIComplete
