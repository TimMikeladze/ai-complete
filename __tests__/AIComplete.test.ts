import { AIComplete } from '../src'

describe('AIComplete', () => {
  it('createCompletion', async () => {
    const aic = new AIComplete({
      openAI: {
        config: {
          apiKey: process.env.OPENAI_API_KEY
        }
      }
    })
    const result = await aic.createCompletion({
      globby: {
        patterns: ['example/locales/en/**/*']
      },
      input: async (args) => ({
        prompt:
          'Translate the JSON below into Russian but keep names of all keys and metadata in English.',
        request: {
          temperature: 0
        }
      }),
      output: async ({ data }) => ({
        choice: JSON.parse(data.choices[0].text)
      })
    })

    expect(result.map((x) => x.output.choice)).toMatchSnapshot()
  })

  it('createEdit', async () => {
    const aic = new AIComplete({
      openAI: {
        config: {
          apiKey: process.env.OPENAI_API_KEY
        }
      }
    })
    const result = await aic.createEdit({
      data: [
        {
          type: 'text',
          value: 'One, two, ___, four.'
        }
      ],
      input: async () => ({
        instruction: 'Fill in the blank with the correct word.'
      }),
      output: async ({ data }) => {
        return {
          choice: data.choices[0].text
        }
      }
    })

    expect(result.map((x) => x.output)).toMatchSnapshot()
  })
})
