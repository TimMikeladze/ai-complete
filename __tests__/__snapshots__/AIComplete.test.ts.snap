// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AIComplete createCompletion 1`] = `
[
  {
    "CareerManagement": {
      "careerPath": "Карьерный путь",
      "description": "{{ count }} найдено",
      "jobCategory": "Категория должности",
      "jobTitle": "Название должности",
      "name": "Career paths",
      "searchToolbarRenderer": {
        "careerPaths": {
          "label": "Карьерные пути",
          "tooltip": "",
        },
        "jobCategories": {
          "label": "Категории должностей",
          "tooltip": "",
        },
        "jobTitles": {
          "label": "Названия должностей",
          "tooltip": "",
        },
      },
      "upload": "Загрузить CSV",
    },
  },
  {
    "UserManagement": {
      "description": "{{ count }} найдено",
      "location": "Местоположение",
      "name": "Управление пользователями",
      "searchToolbarRenderer": {
        "inviteLink": {
          "label": "Пригласить ссылку",
          "tooltip": "",
        },
        "locations": {
          "label": "Местоположения",
          "tooltip": "",
        },
        "roles": {
          "label": "Роли",
          "tooltip": "",
        },
      },
      "upload": "Загрузить CSV",
      "user": "Пользователь",
    },
  },
]
`;

exports[`AIComplete createEdit 1`] = `
[
  {
    "choice": "One, two, three, four.
",
  },
]
`;
